rfc/rfc-0000-multiple-paths.md

```diff
@@ -0,0 +1,241 @@
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSupplyChain
+metadata:
+  name: supply-chain
+spec:
+  resources:
+    - name: source-provider
+      templateRef:
+        kind: ClusterSourceTemplate
+        options:
+          - name: source-from-git-repo
+            selector:
+              matchFields:
+                { key: "spec.source.url", operation: exists }
+          - name: source-from-image-registry
+            selector:
+              matchFields:
+                { key: "spec.source.image", operation: exists }
+
+    - name: source-tester
+      templateRef:
+        kind: ClusterSourceTemplate
+        name: test-source-with-tekton
+      selector:
+        matchLabels:
+          has-tests: "true"
+      sources:
+        - resource: source-provider
+          name: source
+
+    - name: image-provider
+      templateRef:
+        kind: ClusterImageTemplate
+        name: image-from-image-registry
+      selector:
+        matchFields:
+          { key: "spec.image", operation: exists }
+
+    - name: image-builder
+      templateRef:
+        kind: ClusterImageTemplate
+        options:
+          - name: build-image
+            images:
+              - resource: source-tester
+                name: source
+          - name: build-image
+            images:
+              - resource: source-provider
+                name: source
+
+    - name: configure
+      templateRef:
+        kind: ClusterConfigTemplate
+        options:
+          - name: configure
+            sources:
+              - resource: image-builder
+                name: image
+          - name: configure
+            sources:
+              - resource: image-provider
+                name: image
+
+    - name: gitops
+      templateRef:
+        kind: ClusterSourceTemplate
+        options:
+          - name: git-pusher
+            selector:
+              matchLabels:
+                target: gitops
+            configs:
+              - resource: configure
+                name: config
+          - name: registry-pusher
+            selector:
+              matchLabels:
+                target: repostiory
+            configs:
+              - resource: configure
+                name: config
+```
+
+If label `has-tests` is "true", then `image-builder` chooses the first option, from source-tester, because source-tester _will_ eventually be stamped out. 
+Otherwise `image-builder` selects the second option, from source-provider.
+
+This lets a resource be skipped if none of its options match.
+
+`configure` shows a similar pathing, based this time on matchFields not matchLabels.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+An option is selected for one of these reasons:
+
+> It's the only option and all other criteria are met.
+
+OR
+
+> It matches on labels and fields (if no selector is specified, that is a match)
+
+AND
+
+> It's inputs are fulfilled using these same rules.
```

Comment:
I would call this a transitive condition. It is certainly determinitive. But it is difficult to reason about. It seems a reasonable goal that a user looking at a supply-chain can easily determine how a workload would flow through it. Needing to notate choices at earlier nodes, and use those to reason about what choice will be made at a new node is tricky.

This can be alleviated in a few ways.

1. Remove transitive conditions.
2. Require that there is a visualization tool for supply chains. (Fortunately, Rash has already prototyped one!) Such a tool can make implicit logic in the supply-chain explicit in the visualization.
3. Require a mutating webhook that would make all of the implicit transitive conditions explicit conditions at later steps. For example, taking the `is-green` and `is-red` example:

original:
```yaml
kind: ClusterSupplyChain
spec:
  resources:
    - name: some-red-step
      templateRef:
        kind: ClusterSourceTemplate
        name: some-red-template
      selector:
        matchFields:
          { key: "metadata.labels.is-red", operation: in, values: ["true"] }

    - name: some-green-step
      templateRef:
        kind: ClusterSourceTemplate
        name: some-green-template
      selector:
        matchFields:
          { key: "metadata.labels.is-green", operation: in, values: ["true"] }

    - name: some-final-step
      templateRef:
        kind: ClusterSourceTemplate
        options:
          - name: option-1
            sources:
              - resource: some-red-step
                name: source
          - name: option-2
            sources:
              - resource: some-green-step
                name: source
```

transformed by a mutating webhook:
```yaml
kind: ClusterSupplyChain
spec:
  resources:
    - name: some-red-step
      templateRef:
        kind: ClusterSourceTemplate
        name: some-red-template
      selector:
        matchFields:
          { key: "metadata.labels.is-red", operation: in, values: ["true"] }

    - name: some-green-step
      templateRef:
        kind: ClusterSourceTemplate
        name: some-green-template
      selector:
        matchFields:
          { key: "metadata.labels.is-green", operation: in, values: ["true"] }

    - name: some-final-step
      templateRef:
        kind: ClusterSourceTemplate
        options:
          - name: option-1
            sources:
              - resource: some-red-step
                name: source
            selector:        # <--- This field can be added by a mutating webhook
              matchFields:
                { key: "metadata.labels.is-red", operation: in, values: ["true"] }
          - name: option-2
            sources:
              - resource: some-green-step
                name: source
            selector:        # <--- This field can be added by a mutating webhook
              matchFields:
                { key: "metadata.labels.is-green", operation: in, values: ["true"] }
```

Reply:
For 2., wrt vizualization. We don't even need that, just syntax matching that lets you see/jump to definitions.

I don't think it's a particularily heavy ask to suggest that an input matching upstream is a necessity for an option to play out. Transitive dependencies tend to encapsulate their purpose, eg: I (a resource) depend on the source provider, so clearly I wont run if there is no source. I might need to add an `option` to depend on the image-provider, or I might realize I cannot do anything with images, so I do not depend on it. Somewhat OO in it's design, this implies that an upstream resource knows precisely what's needed for it to exist, and you do not need to go prying into it's affairs to know wether you need it's output or not.

Reply:
I think all of that is correct. I was bringing up the user experience of reading a supply chain and trying to reason through the path a workload would take. I suspect that's a reasonable goal.

Reply:
Which is precisely the picture I'm trying to paint. There are trade offs in complex systems between top down and bottom up comprehension. Being able to look at a single component (bottom up) and know why/how it was selected is simply a matter of "what it selects for", not "also what it's predecessors selected for". I don't look at a flow chart and expect to see each decision diamond to contain all the previous choices made. That's when I switch to top down (I look at the whole chart). 

Top down comprehension **is** degraded by multi-pathing, but not by the nature of this RFC. It's the nature of complex dependencies presented in lists/recordsets, especially those with fan-in behavior. If not for multiple input nodes (fan in) we'd be able to use indentation and nest objects under each other, giving us a wonderful visual clue in the document structure, but we don't have that luxury. 

Denormalizing selectors (as proposed above) does not help with visualizing paths either, it is also a bottom-up solution.

I'd rather we find clear ways of annotating paths through a supply chain. We can (if we uphold determinism) annotate a workload (in the status!) with a reduced list of resources from the chosen path (just like RFC 20's artifact mapping). 

Even without multi-pathing, visualization and 'go to definition' tooling is a valuable addition, so we should *ensure we do that*.